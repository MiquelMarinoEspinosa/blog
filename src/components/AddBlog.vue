<template>
  <div id="add-blog">
    <h2>Add a New Blog Post</h2>
    <form v-if="!submitted">
      <label>Blog Title:</label>
      <input type="text" required v-model="blog.title" />
      <label>Blog Content:</label>
      <textarea v-model="blog.content"></textarea>

      <div id="checkboxes">
        <label>Vue</label>
        <input type="checkbox" value="vue" v-model="blog.categories" />
        <label>React</label>
        <input type="checkbox" value="react" v-model="blog.categories" />
        <label>Angular</label>
        <input type="checkbox" value="angular" v-model="blog.categories" />
        <label>Nodejs</label>
        <input type="checkbox" value="nodejs" v-model="blog.categories" />
      </div>
      <label>Author:</label>
      <select v-model="blog.author">
        <option v-for="(author, index) in authors" :key="index">
          {{ author }}
        </option>
      </select>
      <button @click.prevent="post">Add Blog</button>
    </form>
    <div v-if="submitted">
      <h3>Thanks for adding your post</h3>
    </div>
    <div id="preview">
      <h3>Preview Blog</h3>
      <p>Blog title: {{ blog.title }}</p>
      <p>Blog content:</p>
      <p>{{ blog.content }}</p>
      <p>Blog categories:</p>
      <ul>
        <li v-for="(category, index) in blog.categories" :key="index">
          {{ category }}
        </li>
      </ul>
      <p>Author: {{ blog.author }}</p>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      blog: {
        title: "",
        content: "",
        categories: [],
        author: "",
      },
      authors: ["Luke Winslow", "Mike Fernandez", "Don Hattaway"],
      submitted: false,
    };
  },
  methods: {
    post() {
      this.$http.post("http://localhost:8081/blog", this.blog).then(() => {
        this.submitted = true;
      });
    },
  },
};
</script>

<style>
#add-blog * {
  box-sizing: border-box;
}
#add-blog {
  margin: 20px auto;
  max-width: 500px;
}
label {
  display: block;
  margin: 20px 0 10px;
}
input[type="text"],
textarea {
  display: block;
  width: 100%;
  padding: 8px;
}
#preview {
  padding: 10px 20px;
  border: 1px dotted #ccc;
  margin: 30px 0;
}
h3 {
  margin-top: 10px;
}
#checkboxes input {
  display: inline-block;
  margin-right: 10px;
}
#checkboxes label {
  display: inline-block;
}
</style>
