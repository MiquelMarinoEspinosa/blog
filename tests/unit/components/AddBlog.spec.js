/* 
  Improvements:
    - focus the search subelements within elements (blog title inside form) using searching
    - change the input's text values using the value attribute
    - use await for changin events like mouse clicks or blur
*/

import { mount } from '@vue/test-utils'
import AddBlog from '@/components/AddBlog.vue'

describe('AddBlog', () => {
  const wrapper = mount(AddBlog)

  it('contains the main div id "add-blog"', () => {
    expect(wrapper.html()).toContain('<div id="add-blog">')
  })

  it('contains the title "Add a New Blog Post"', () => {
    expect(wrapper.html()).toContain('<h2>Add a New Blog Post</h2>')
  })

  it('contains the form section', () => {
    expect(wrapper.html()).toContain('<form>')
    expect(wrapper.html()).toContain('</form>')
  })

  it('form section contains the "Blog Title"', () => {
    expect(wrapper.html()).toContain(
      '<label>Blog Title:</label> <input type="text" required="required">'
    )
  })

  it('form section contains the "Blog Content" label', () => {
    expect(wrapper.html()).toContain(
      '<label>Blog Content:</label> <textarea></textarea>'
    )
  })

  it('contains the checkboxes section', () => {
    expect(wrapper.html()).toContain('<div id="checkboxes">')
  })

  it('checkboxes section contains "Ninjas" option', () => {
    expect(wrapper.html()).toContain(
      '<label>Ninjas</label> <input type="checkbox" value="ninjas">'
    )
  })

  it('checkboxes section contains "Wizards" option', () => {
    expect(wrapper.html()).toContain(
      '<label>Wizards</label> <input type="checkbox" value="wizards">'
    )
  })

  it('checkboxes section contains "Mario" option', () => {
    expect(wrapper.html()).toContain(
      '<label>Mario</label> <input type="checkbox" value="mario">'
    )
  })

  it('checkboxes section contains "Cheese" option', () => {
    expect(wrapper.html()).toContain(
      '<label>Cheese</label> <input type="checkbox" value="cheese">'
    )
  })

  it('contains the preview section', () => {
    expect(wrapper.html()).toContain('<div id="preview">')
  })

  it('preview section contains the title "Preview Blog"', () => {
    expect(wrapper.html()).toContain('<h3>Preview Blog</h3>');
  })

  it('preview section contains the "Blog title"', () => {
    expect(wrapper.html().toContain(
      '<p>Blog title: </p>'
    );
  })

  it('preview section contains the "Blog content"', () => {
    expect(wrapper.html()).toContain(
      '<p>Blog content:</p>\n    <p></p>'
    );
  })

  it('preview section contians the "Blog categories"', () => {
    expect(wrapper.html()).toContain(
      '<p>Blog categories:</p>'
    );
  })
});
