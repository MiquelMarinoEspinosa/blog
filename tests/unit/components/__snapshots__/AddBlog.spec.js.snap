// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBlog initial render and creating a blog entry fills out the form fields and updates the preview section 1`] = `
<div id="add-blog">
  <h2>Add a New Blog Post</h2>
  <form><label>Blog Title:</label> <input type="text" required="required"> <label>Blog Content:</label> <textarea></textarea>
    <div id="checkboxes"><label>Vue</label> <input type="checkbox" value="vue"> <label>React</label> <input type="checkbox" value="react"> <label>Angular</label> <input type="checkbox" value="angular"> <label>Nodejs</label> <input type="checkbox" value="nodejs"></div> <label>Author:</label> <select>
      <option>Author 1</option>
      <option>Author 2</option>
      <option>Author 3</option>
    </select> <button>Add Blog</button>
  </form>
  <!---->
  <div id="preview">
    <h3>Preview Blog</h3>
    <p>Blog title: New Blog</p>
    <p>Blog content:</p>
    <p>New Blog Content</p>
    <p>Blog categories:</p>
    <ul>
      <li>vue</li>
    </ul>
    <p>Author: Author 1</p>
  </div>
</div>
`;

exports[`AddBlog initial render and creating a blog entry renders the initial html 1`] = `
<div id="add-blog">
  <h2>Add a New Blog Post</h2>
  <form><label>Blog Title:</label> <input type="text" required="required"> <label>Blog Content:</label> <textarea></textarea>
    <div id="checkboxes"><label>Vue</label> <input type="checkbox" value="vue"> <label>React</label> <input type="checkbox" value="react"> <label>Angular</label> <input type="checkbox" value="angular"> <label>Nodejs</label> <input type="checkbox" value="nodejs"></div> <label>Author:</label> <select>
      <option>Author 1</option>
      <option>Author 2</option>
      <option>Author 3</option>
    </select> <button>Add Blog</button>
  </form>
  <!---->
  <div id="preview">
    <h3>Preview Blog</h3>
    <p>Blog title: </p>
    <p>Blog content:</p>
    <p></p>
    <p>Blog categories:</p>
    <ul></ul>
    <p>Author: </p>
  </div>
</div>
`;

exports[`AddBlog initial render and creating a blog entry send the form and create the blog entry 1`] = `
<div id="add-blog">
  <h2>Add a New Blog Post</h2>
  <!---->
  <div>
    <h3>Thanks for adding your post</h3>
  </div>
  <div id="preview">
    <h3>Preview Blog</h3>
    <p>Blog title: New Blog</p>
    <p>Blog content:</p>
    <p>New Blog Content</p>
    <p>Blog categories:</p>
    <ul>
      <li>vue</li>
    </ul>
    <p>Author: Author 1</p>
  </div>
</div>
`;
